### PR TITLE
use pytest for cpp tests

### LINKED ISSUE #29827: use pytest for cpp tests
https://github.com/pytest-dev/pytest-cpp

### PR DESCRIPTION
**Pull request by adeebshihadeh**

### CODE REVIEW

File: common/tests/test_util.cc [hunk 1]
```diff
@@ -38,7 +42,7 @@ TEST_CASE("util::read_file") {
     std::string content = random_bytes(64 * 1024);
     write(fd, content.c_str(), content.size());
     std::string ret = util::read_file(filename);
-    REQUIRE(ret == content);
+    REQUIRE(hasher(ret) == hasher(content));
```
Comment (jnewb1) [inline common/tests/test_util.cc:45]: asserting the whole files gets dumped into the XML file, easier to just compare the hashes

File: system/loggerd/tests/test_logger.cc [hunk 2]
```diff
@@ -97,7 +97,6 @@ TEST_CASE("logger") {
 
     auto logging_thread = [&]() -> void {
       LoggerHandle *lh = logger_get_handle(&logger);
-      REQUIRE(lh != nullptr);
```
Comment (jnewb1) [inline system/loggerd/tests/test_logger.cc:100]: catch2 is not thread safe

**Comment by jnewb1** [comment]
pytest-cpp doesn't properly support the SECTION stuff within catch2. I made an upstream PR to fix that: https://github.com/pytest-dev/pytest-cpp/pull/112

File: system/proclogd/tests/test_proclog.cc [hunk 3]
```diff
@@ -140,7 +140,6 @@ TEST_CASE("buildProcLogerMessage") {
       REQUIRE(p.getName() == "test_proclog");
       REQUIRE(p.getState() == 'R');
       REQUIRE_THAT(p.getExe().cStr(), Catch::Matchers::Contains("test_proclog"));
-      REQUIRE(p.getCmdline().size() == 1);
```
Comment (jnewb1) [inline system/proclogd/tests/test_proclog.cc:143]: test has more than 1 argument when running with pytest-cpp

File: pyproject.toml [hunk 4]
```diff
@@ -142,6 +146,7 @@ pygame = "*"
 pyprof2calltree = "*"
 pytest = "*"
 pytest-cov = "*"
+pytest-cpp = { git = "https://github.com/jnewb1/pytest-cpp.git", branch = "catch2-fix-section" } # until it's merged upstream
```
Comment (jnewb1) [inline pyproject.toml:149]: should not merge this till its upstreamed

**Comment by jnewb1** [comment]
This also makes it way easier to see the times for each C++ test.

test_replay is almost 2 minutes
'''107.05s call     tools/replay/tests/test_replay::Route'''

**Comment by jnewb1** [comment]
@deanlee 

On tests/test_cabana and tests/test_replay, it will fail if the route isn't already downloaded because you need a JWT to download from the comma servers. This wasn't an issue in CI before because the route is downloaded in other tests before those are run. You can replicate this locally by deleting /tmp/comma_download_cache, or using an openpilot_prefix

**Comment by deanlee** [comment]
> On tests/test_cabana and tests/test_replay, it will fail if the route isn't already...

This is due to the original TEST_RLOG_URL is no longer public:
https://github.com/commaai/openpilot/blob/fb1b1f15aaaf401b20f790e61aebeb1b6d2c9249/tools/cabana/tests/test_cabana.cc#L10

I'll fix it

**Comment by jnewb1** [comment]
it says it's public? but I can't download it without a token
image

**Comment by deanlee** [comment]
the TEST_RLOG_URL in test_replay.cc is still publicly accessible. https://commadataci.blob.core.windows.net/openpilotci/0c94aa1e1296d7c6/2021-05-05--19-48-37/0/rlog.bz2

File: system/loggerd/tests/test_logger.cc (hunk 2 above)
Comment (deanlee) [reply system/loggerd/tests/test_logger.cc:100]: fixed: https://github.com/commaai/openpilot/pull/29852

File: common/tests/test_util.cc [hunk 5]
```diff
@@ -9,9 +9,13 @@
 #include <random>
 #include <string>
 
+#include <unordered_map>
+
 #include "catch2/catch.hpp"
 #include "common/util.h"
 
+std::hash<std::string> hasher;
```
Comment (deanlee) [inline common/tests/test_util.cc:17]: this line can be removed

File: common/tests/test_util.cc (hunk 1 above)
Comment (jnewb1) [reply common/tests/test_util.cc]: done in another PR

File: common/tests/test_util.cc (hunk 5 above)
Comment (jnewb1) [reply common/tests/test_util.cc]: done

**Comment by jnewb1** [comment]
this should be good to review now. Moved cpp_files to CI options until we update agnos.

**Review by adeebshihadeh — COMMENTED** [review]
LGTM

**Pull request merged**